src/asfvideo.cpp

```diff
@@ -0,0 +1,770 @@
+  localPosition_ = io_->tell();
+}  // AsfVideo::decodeBlock
+
+void AsfVideo::extendedStreamProperties(uint64_t size) {
+  uint64_t cur_pos = io_->tell(), avgTimePerFrame = 0;
+  DataBuf buf(8);
+  static int previousStream;
+  io_->seek(cur_pos + 48, BasicIo::beg);
+
+  std::memset(buf.data(), 0x0, buf.size());
+  io_->read(buf.data(), 2);
+  streamNumber_ = Exiv2::getUShort(buf.data(), littleEndian);
+
+  io_->read(buf.data(), 2);
+  io_->read(buf.data(), 8);
+  avgTimePerFrame = getUint64_t(buf);
+
+  if (previousStream < streamNumber_ && avgTimePerFrame != 0)
+    xmpData()["Xmp.video.FrameRate"] = (double)10000000 / (double)avgTimePerFrame;
+
+  previousStream = streamNumber_;
+  io_->seek(cur_pos + size, BasicIo::beg);
+}  // AsfVideo::extendedStreamProperties
+
+void AsfVideo::contentDescription(uint64_t size) {
+  const size_t pos = io_->tell();
+  size_t length[5];
+  for (int i = 0; i < 5; ++i) {
+    byte buf[2];
+    io_->read(buf, 2);
+    if (io_->error() || io_->eof())
+      throw Error(ErrorCode::kerFailedToReadImageData);
+    length[i] = getUShort(buf, littleEndian);
+  }
+  for (int i = 0; i < 5; ++i) {
+    DataBuf buf(length[i]);
+    std::memset(buf.data(), 0x0, buf.size());
+    io_->read(buf.data(), length[i]);
+    if (io_->error() || io_->eof())
+      throw Error(ErrorCode::kerFailedToReadImageData);
+    const TagDetails* td = find(contentDescriptionTags, i);
+    assert(td);
+    std::string str((const char*)buf.data(), length[i]);
+    if (convertStringCharset(str, "UCS-2LE", "UTF-8")) {
+      xmpData()[td->label_] = str;
+    } else {
+      xmpData()[td->label_] = toString16(buf);
+    }
+  }
+  if (io_->seek(pos + size, BasicIo::beg))
+    throw Error(ErrorCode::kerFailedToReadImageData);
+}  // AsfVideo::contentDescription
+
+void AsfVideo::streamProperties() {
+  DataBuf buf(20);
+  buf.data()[8] = '\0';
+  byte guidBuf[16];
+  int stream = 0;
+  io_->read(guidBuf, 16);
+  char streamType[37] = "";
+  // Exiv2::RiffVideo *test = NULL; todo
+
+  getGUID(guidBuf, streamType);
+  const TagVocabulary* tv;
+  tv = find(GUIDReferenceTags, streamType);
+  io_->read(guidBuf, 16);
+
+  if (compareTag(exvGettext(tv->label_), "Audio_Media"))
+    stream = 1;
+  else if (compareTag(exvGettext(tv->label_), "Video_Media"))
+    stream = 2;
+
+  io_->read(buf.data(), 8);
+  if (stream == 2)
+    xmpData()["Xmp.video.TimeOffset"] = getUint64_t(buf);
+  else if (stream == 1)
+    xmpData()["Xmp.audio.TimeOffset"] = getUint64_t(buf);
+
+  io_->read(buf.data(), 8);
+  std::memset(buf.data(), 0x0, buf.size());
+  io_->read(buf.data(), 1);
+  streamNumber_ = (int)buf.data()[0] & 127;
+
+  io_->read(buf.data(), 5);
+  std::memset(buf.data(), 0x0, buf.size());
+  io_->read(buf.data(), 2);
+  size_t temp = Exiv2::getUShort(buf.data(), littleEndian);
+
+  if (stream == 2) {
+    xmpData()["Xmp.video.Width"] = temp;
+    width_ = temp;
+  } else if (stream == 1) {
+    // xmpData()["Xmp.audio.Codec"] = test->printAudioEncoding(temp);
```

Comment:
The whole `else if` branch is now empty. I would just remove it or leave a TODO comment in case you think we should that care about that `stream == 1`. By the way, instead of using the magic values `1` and `2` it would be useful to create a enum to give names to these values. I noticed that they are used in more places and it would improve the code readability.

Reply:
Done here ad72e1e6b0c2a10383a3878c1e603aa78fe3275a